framework/src/main/java/org/checkerframework/framework/util/ComplexQualifierHierarchy.java

```diff
@@ -0,0 +1,375 @@
+            for (AnnotationMirror superAnno : superAnnos) {
+                QualifierKind superKind = getQualifierKind(superAnno);
+                if (subKind.areInSameHierarchy(superKind)) {
+                    if (isSubtype(subAnno, superAnno)) {
+                        isSubtypeCount++;
+                    } else {
+                        return false;
+                    }
+                }
+            }
+        }
+        return isSubtypeCount == superAnnos.size();
+    }
+
+    @Override
+    public AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
+        QualifierKind qual1 = getQualifierKind(a1);
+        QualifierKind qual2 = getQualifierKind(a2);
+        if (!qual1.areInSameHierarchy(qual2)) {
+            return null;
+        }
+        QualifierKind lub = qualifierKindHierarchy.leastUpperBound(qual1, qual2);
+        if (lub.hasElements()) {
+            return leastUpperBound(a1, qual1, a2, qual2);
+        }
+        return qualifierMap.get(lub);
+    }
+
+    /**
+     * Returns the least upper bound between {@code a1} and {@code a2}.
+     *
+     * <p>This method is only called when the lub is an annotation with elements.
+     *
+     * @param a1 first annotation
+     * @param qual1 QualifierKind for {@code a1}
+     * @param a2 second annotation
+     * @param qual2 QualifierKind for {@code a2}
+     * @return the least upper bound between {@code a1} and {@code a2}.
+     */
+    protected abstract AnnotationMirror leastUpperBound(
+            AnnotationMirror a1, QualifierKind qual1, AnnotationMirror a2, QualifierKind qual2);
+
+    @Override
+    public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
+        QualifierKind qual1 = getQualifierKind(a1);
+        QualifierKind qual2 = getQualifierKind(a2);
+        if (!qual1.areInSameHierarchy(qual2)) {
+            return null;
+        }
+        QualifierKind glb = qualifierKindHierarchy.greatestLowerBound(qual1, qual2);
+        if (glb.hasElements()) {
+            return greatestLowerBound(a1, qual1, a2, qual2);
+        }
+        return qualifierMap.get(glb);
+    }
+
+    /**
+     * Returns the greatest lower bound between {@code a1} and {@code a2}.
+     *
+     * <p>This method is only called when the glb is an annotation with elements.
+     *
+     * @param a1 first annotation
+     * @param qual1 QualifierKind for {@code a1}
+     * @param a2 second annotation
+     * @param qual2 QualifierKind for {@code a2}
+     * @return the greatest lower bound between {@code a1} and {@code a2}.
+     */
+    protected abstract AnnotationMirror greatestLowerBound(
+            AnnotationMirror a1, QualifierKind qual1, AnnotationMirror a2, QualifierKind qual2);
+
+    @Override
+    public boolean isSubtypeTypeVariable(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+        if (superAnno == null) {
+            // [] is a supertype of any qualifier, and [] <: []
```

Comment:
I don't understand what "[]" means in this comment (and in other similar comments).